clang/lib/CIR/Dialect/IR/CIRTypes.cpp

```diff
@@ -1010,45 +995,52 @@ static void printFuncTypeArgs(mlir::AsmPrinter &p,
+                                       mlir::Type &optionalReturnType,
                                        llvm::SmallVector<mlir::Type> &params,
                                        bool &isVarArg) {
-  if (failed(parseFuncTypeReturn(p, optionalReturnTypes)))
+  if (failed(parseFuncTypeArgs(p, params, isVarArg)))
     return failure();
-  return parseFuncTypeArgs(p, params, isVarArg);
+  return parseFuncTypeReturn(p, optionalReturnType);
 }
 
-static void printFuncType(mlir::AsmPrinter &p, mlir::Type optionalReturnTypes,
+static void printFuncType(mlir::AsmPrinter &p, mlir::Type optionalReturnType,
                           mlir::ArrayRef<mlir::Type> params, bool isVarArg) {
-  printFuncTypeReturn(p, optionalReturnTypes);
   printFuncTypeArgs(p, params, isVarArg);
+  printFuncTypeReturn(p, optionalReturnType);
 }
 
-// Return the actual return type or an explicit !cir.void if the function does
-// not return anything
+/// Get the C-style return type of the function, which is !cir.void if the
+/// function returns nothing and the actual return type otherwise.
 mlir::Type FuncType::getReturnType() const {
-  if (isVoid())
+  if (hasVoidReturn())
     return cir::VoidType::get(getContext());
-  return static_cast<detail::FuncTypeStorage *>(getImpl())->optionalReturnType;
+  return getOptionalReturnType();
 }
 
-/// Returns the result type of the function as an ArrayRef, enabling better
-/// integration with generic MLIR utilities.
+/// Get the MLIR-style return type of the function, which is an empty
+/// ArrayRef if the function returns nothing and a single-element ArrayRef
+/// with the actual return type otherwise.
 llvm::ArrayRef<mlir::Type> FuncType::getReturnTypes() const {
-  if (isVoid())
+  if (hasVoidReturn())
     return {};
-  return static_cast<detail::FuncTypeStorage *>(getImpl())->optionalReturnType;
-}
-
-// Whether the function returns void
-bool FuncType::isVoid() const {
-  auto rt =
-      static_cast<detail::FuncTypeStorage *>(getImpl())->optionalReturnType;
-  assert(!rt ||
-         !mlir::isa<cir::VoidType>(rt) &&
-             "The return type for a function returning void should be empty "
-             "instead of a real !cir.void");
-  return !rt;
+  // Can't use getOptionalReturnType() here because llvm::ArrayRef hold a
+  // pointer to its elements and doesn't do lifetime extension.  That would
+  // result in returning a pointer to a temporary that has gone out of scope.
+  return getImpl()->optionalReturnType;
+}
+
+// Does the fuction type return nothing?
+bool FuncType::hasVoidReturn() const { return !getOptionalReturnType(); }
+
+mlir::LogicalResult
+FuncType::verify(llvm::function_ref<mlir::InFlightDiagnostic()> emitError,
+                 llvm::ArrayRef<mlir::Type> argTypes, mlir::Type returnType,
+                 bool isVarArg) {
+  if (returnType && mlir::isa<cir::VoidType>(returnType)) {
+    emitError() << "!cir.func cannot have an explicit 'void' return type";
```

Comment:
We need to add a test for this one, perhaps in `invalid.cir`?

Reply:
Done.  I added a new test case at the end of `invalid.cir`.